cpp/src/join/join_common_utils.cuh

```diff
@@ -0,0 +1,263 @@
+{
+  // Get array of indices that do not appear in right_indices
+
+  // Vector allocated for unmatched result
+  auto right_indices_complement =
+    std::make_unique<rmm::device_uvector<size_type>>(right_table_row_count, stream);
+
+  // If left table is empty in a full join call then all rows of the right table
+  // should be represented in the joined indices. This is an optimization since
+  // if left table is empty and full join is called all the elements in
+  // right_indices will be JoinNoneValue, i.e. -1. This if path should
+  // produce exactly the same result as the else path but will be faster.
+  if (left_table_row_count == 0) {
+    thrust::sequence(rmm::exec_policy(stream),
+                     right_indices_complement->begin(),
+                     right_indices_complement->end(),
+                     0);
+  } else {
+    // Assume all the indices in invalid_index_map are invalid
+    auto invalid_index_map =
+      std::make_unique<rmm::device_uvector<size_type>>(right_table_row_count, stream);
+    thrust::uninitialized_fill(
+      rmm::exec_policy(stream), invalid_index_map->begin(), invalid_index_map->end(), int32_t{1});
+
+    // Functor to check for index validity since left joins can create invalid indices
+    valid_range<size_type> valid(0, right_table_row_count);
+
+    // invalid_index_map[index_ptr[i]] = 0 for i = 0 to right_table_row_count
+    // Thus specifying that those locations are valid
+    thrust::scatter_if(rmm::exec_policy(stream),
+                       thrust::make_constant_iterator(0),
+                       thrust::make_constant_iterator(0) + right_indices->size(),
+                       right_indices->begin(),      // Index locations
+                       right_indices->begin(),      // Stencil - Check if index location is valid
+                       invalid_index_map->begin(),  // Output indices
+                       valid);                      // Stencil Predicate
+    size_type begin_counter = static_cast<size_type>(0);
+    size_type end_counter   = static_cast<size_type>(right_table_row_count);
+
+    // Create list of indices that have been marked as invalid
+    size_type indices_count = thrust::copy_if(rmm::exec_policy(stream),
+                                              thrust::make_counting_iterator(begin_counter),
+                                              thrust::make_counting_iterator(end_counter),
+                                              invalid_index_map->begin(),
+                                              right_indices_complement->begin(),
+                                              thrust::identity<size_type>()) -
+                              right_indices_complement->begin();
+    right_indices_complement->resize(indices_count, stream);
+  }
+
+  auto left_invalid_indices =
+    std::make_unique<rmm::device_uvector<size_type>>(right_indices_complement->size(), stream);
+  thrust::fill(rmm::exec_policy(stream),
```

Comment:
Why is this `fill` but the other is `uninitialized_fill`?

Reply:
Both hash joins and nested loop (conditional) joins perform full joins by first doing a left join and then going back and filling in pairs corresponding to rows in the right table that have no match in the left table. This functions is responsible for calculating those remaining pairs. All of those pairs are of the form `(JoinNoneValue, RIGHT_INDEX)`. We need to use `thrust::fill` here to create a vector filled with `JoinNoneValue`, whereas for the `right_invalid_indices` we can use `uninitialized_fill` because we anyway need to [launch another kernel](https://github.com/rapidsai/cudf/pull/8815/files#diff-651159b8afa9d590686790b6f73d9fcd0d230bed8a710efa66d96b1d00a6125dR175) to actually populate the list of right indices that have no matching row in the left table.

Reply:
That is a good question. I think both can probably use `fill`.